train/deeplab_train.py

```diff
@@ -275,33 +275,61 @@ def get_random_training_window(raster_ds, label_ds, width, height, window_size,
 
     # We assume here that the ND value will be on the first band
     if img_nd is not None:
-        img_nds = raster_ds.read(1, window=window) == img_nd
+        a = raster_ds.read(1, window=window)
+        img_nds = (a == img_nd) + (np.isnan(a))
     else:
         img_nds = np.zeros(labels.shape)
 
     # nodata mask for regions without labels
     nodata = (img_nds + label_nds) > 0
-    not_nodata = (nodata == 0)
 
     # Normalized float32 imagery bands
     data = []
     for band in bands:
         a = raster_ds.read(band, window=window)
         a = np.array((a - MEANS[band-1]) / STDS[band-1], dtype=np.float32)
-        a = a * not_nodata
+        a[nodata != 0] = 0.0
         data.append(a)
     data = np.stack(data, axis=0)
 
     return (data, labels)
 
 
+def _get_random_training_window(n):
+    return get_random_training_window(Raster_ds, Label_ds,
+                                      Width, Height, Window_size, Bands,
+                                      Label_mappings, Label_nd, Img_nd)
+
 def get_random_training_batch(raster_ds, label_ds, width, height, window_size, batch_size, device, bands, label_mappings, label_nd, img_nd):
     data = []
     labels = []
-    for i in range(0, batch_size):
-        d, l = get_random_training_window(raster_ds, label_ds, width, height, window_size, bands, label_mappings, label_nd, img_nd)
-        data.append(d)
-        labels.append(l)
+
+    global Raster_ds
+    Raster_ds = raster_ds
+    global Label_ds
+    Label_ds = label_ds
+    global Width
+    Width = width
+    global Height
+    Height = height
+    global Window_size
+    Window_size = window_size
+    global Bands
+    Bands = bands
+    global Label_mappings
+    Label_mappings = label_mappings
+    global Label_nd
+    Label_nd = label_nd
+    global Img_nd
+    Img_nd = img_nd
+
+    with Pool(32) as p:
+        for d, l in p.map(_get_random_training_window, range(0, batch_size)):
```

Comment:
This is happening in a different *process*, so hopefully there is no GDAL-related thread-safety issue (hopefully).

Reply:
This is worth a shot. I'm kind of skeptical that it won't blow up in our faces, but the fix is likely just creating a new `dataset` when reads become necessary